transactor/src/main/scala/io/mediachain/transactor/Copycat.scala

```diff
@@ -44,32 +48,204 @@ object Copycat {
+
+  sealed abstract class ClientState
+  
+  object ClientState {
+    case object Connected extends ClientState
+    case object Suspended extends ClientState
+    case object Disconnected extends ClientState
+  }
+
+  trait ClientStateListener {
+    def onStateChange(state: ClientState): Unit
+  }
   
   class Client(client: CopycatClient) extends JournalClient {
+    import scala.concurrent.ExecutionContext.Implicits.global
+    import ClientState._
+    
+    private var shutdown = false
+    private var state: ClientState = Disconnected
+    private var server: Option[String] = None
+    private var reconnectThread: Option[Thread] = None
     private var listeners: Set[JournalListener] = Set()
+    private var stateListeners: Set[ClientStateListener] = Set()
+    private val logger = LoggerFactory.getLogger(classOf[Client])
+    private val random = new Random
+    private val timer = new Timer(true) // runAsDaemon
+ 
+    client.onStateChange(new Consumer[CopycatClient.State] {
+                              def accept(state: CopycatClient.State) {
+                                onStateChange(state)
+                              }
+    })
+    
+    // submit a state machine operation with retry logic to account
+    // for potentially transient client connectivity errors
+    val maxRetries = 5
+    private def submit[T](op: Operation[T], retry: Int = 0): Future[T] = {
+      (state, shutdown) match {
+        case (_, true) =>
+          Future {throw new ClientException("Copycat client has shutdown")}
+
+        case (Disconnected, _) => 
+          Future {throw new ClientException("Copycat client is disconnected")}
+
+        case (Suspended, _) =>
+          logger.info("Copycat client is suspended; delaying "+ op)
+          backoff(1).flatMap { _ =>
+            logger.info("Retrying operation " + op)
+            submit(op, retry)
+          }
+          
+        case (Connected, _) =>
+          val fut = FutureConverters.toScala(client.submit(op))
+          if (retry < maxRetries) {
+            fut.recoverWith { 
+              case e: Throwable =>
+                logger.error("Copycat client error in " + op, e)
+                backoff(retry).flatMap { retry => 
+                  logger.info("Retrying operation " + op)
+                  submit(op, retry + 1)
+                }
+            }
+          } else {
+            fut
+          }
+      }
+    }
+    
+    private def backoff(retry: Int): Future[Int] = {
+      val promise = Promise[Int]
+      val delay = random.nextInt(Math.pow(2, retry).toInt * 1000)
+      
+      logger.info("Backing off for " + delay + " ms")
+      timer.schedule(new TimerTask {
+        def run {
+          promise.complete(Try(retry))
+        }
+      }, delay)
+      promise.future
+    }
+    
+    private def onStateChange(cstate: CopycatClient.State) {
+      cstate match {
+        case CopycatClient.State.CONNECTED => 
+          state = Connected
```

Comment:
the timer has a daemon thread, but it doesn't hang -- it passes control to the other threads through the future pipeline by completing the promise.